src/sentry/sentry_apps/installations.py

```diff
@@ -0,0 +1,174 @@
+@dataclasses.dataclass
+class SentryAppInstallationTokenCreator:
+    sentry_app_installation: SentryAppInstallation
+    expires_at: datetime.date | None = None
+    generate_audit: bool = False
+
+    def run(self, user: User, request: Request | None = None) -> ApiToken:
+        with transaction.atomic():
+            self._check_token_limit()
+            api_token = self._create_api_token()
+            self._create_sentry_app_installation_token(api_token=api_token)
+            self.audit(request, api_token)
+        self.record_analytics(user)
+        return api_token
+
+    def _check_token_limit(self) -> None:
+        curr_count = SentryAppInstallationToken.objects.filter(
+            sentry_app_installation=self.sentry_app_installation
+        ).count()
+        if curr_count >= INTERNAL_INTEGRATION_TOKEN_COUNT_MAX:
+            raise ApiTokenLimitError(
+                "Cannot generate more than %d tokens for a single integration"
+                % INTERNAL_INTEGRATION_TOKEN_COUNT_MAX
+            )
+
+    def _create_api_token(self) -> ApiToken:
+        return ApiToken.objects.create(
+            user=self.sentry_app.proxy_user,
+            application_id=self.sentry_app.application.id,
+            scope_list=self.sentry_app.scope_list,
+            expires_at=self.expires_at,
+        )
+
+    def _create_sentry_app_installation_token(
+        self, api_token: ApiToken
+    ) -> SentryAppInstallationToken:
+        return SentryAppInstallationToken.objects.create(
+            api_token=api_token, sentry_app_installation=self.sentry_app_installation
+        )
+
+    def audit(self, request: Request | None, api_token: ApiToken) -> None:
+        from sentry.utils.audit import create_audit_entry
+
+        if request and self.generate_audit:
+            create_audit_entry(
+                request=request,
+                organization=self.organization,
+                target_object=api_token.id,
+                event=audit_log.get_event_id("INTERNAL_INTEGRATION_ADD_TOKEN"),
+                data={"sentry_app": self.sentry_app.name},
+            )
+
+    def record_analytics(self, user: User) -> None:
+        from sentry import analytics
+
+        analytics.record(
+            "sentry_app_installation_token.created",
+            user_id=user.id,
+            organization_id=self.organization.id,
+            sentry_app_installation_id=self.sentry_app_installation.id,
+            sentry_app=self.sentry_app.slug,
+        )
+
+    @cached_property
+    def sentry_app(self) -> SentryApp:
+        return self.sentry_app_installation.sentry_app
+
+    @cached_property
+    def organization(self) -> Organization:
+        return self.sentry_app_installation.organization
+
+
+@dataclasses.dataclass
+class SentryAppInstallationCreator:
+    organization_id: int
+    slug: str
+    notify: bool = True
+
+    def run(self, *, user: User, request: Request | None) -> SentryAppInstallation:
+        with transaction.atomic():
```

Comment:
Should we be use connection selection here? It might help in the future when control silo resources are in a separate database.

Reply:
I hadn't been thinking about doing connection selection explicitly, but rather implicitly.  For instance, all endpoints marked control silo would force all subsequent code to route to control silo set of databases.